src/Services/OpenApiGenerator.php

```diff
@@ -366,4 +315,243 @@ private function getResponses(Route $route): array
 
         return $responses;
     }
+
+    private function buildPropertiesSchema(array $rules): array
+    {
+        $schema = [];
+        foreach ($rules as $key => $value) {
+            if ('#rule' === $key) {
+                continue;
+            }
+            if (is_string($value)) { // Property
+                $schema[$key] = $this->buildSchemaFromRule($value);
+            } else {
+                $schema[$key] = $this->buildSchemaFromRule($value['#rule'] ?? '');
+                if (isset($value['*'])) { // Array
+                    $schema[$key]['type'] = 'array';
+                    if (is_string($value['*'])) { // Array of Values
+                        $items = $this->buildSchemaFromRule($value['*']);
+                    } else { // Array of Objects
+                        $items               = $this->buildSchemaFromRule($value['*']['#rule'] ?? '');
+                        $items['type']       = 'object';
+                        $items['properties'] = $this->buildPropertiesSchema($value['*']);
+                    }
+                    $schema[$key]['items'] = $items;
+                } else { // Object
+                    $schema[$key]['type']       = 'object';
+                    $schema[$key]['properties'] = $this->buildPropertiesSchema($value);
+                }
+            }
+        }
+
+        return $schema;
+    }
+
+    private function buildSchemaFromRule(string $ruleSpec): array
+    {
+        $schema = [
+            'type' => 'string',
+        ];
+
+        if ('' === $ruleSpec) {
+            return $schema;
+        }
+
+        $rules = $this->parseValidationRuleSpec($ruleSpec);
+
+        $ruleHandlers = [];
+        foreach (array_keys($rules) as $rule) {
+            if (isset($this->ruleHandlers[$rule])) {
+                $ruleHandlers[$rule] ??= [];
+                $ruleHandlers[$rule][] = $this->ruleHandlers[$rule];
+            }
+            foreach ($this->ruleHandlers as $rx => $value) {
+                if (!ctype_alpha($rx[0]) && preg_match($rx, $rule)) {
+                    $ruleHandlers[$rule] ??= [];
+                    $ruleHandlers[$rule][] = $value;
+                }
+            }
+        }
+
+        foreach ($ruleHandlers as $rule => $handlers) {
+            foreach ($handlers as $handler) {
+                $schema = $handler($schema, $rule, $rules[$rule], $rules, $this->ruleHandlers);
+            }
+        }
+
+        return array_filter($schema, fn ($v) => null !== $v);
+    }
+
+    private function parseValidationRuleSpec(string $ruleSpec): array
+    {
+        static $cache = [];
```

Comment:
Please avoid the use of local static vars as they cannot be garbage collected when running under php-pm, or cleared by a clear() method (for example). A class property and a clear method would be better (perhaps consider adding the `Symfony\Contracts\Service\ResetInterface`).